### ISSUE TITLE
changed format of trace graph in torch 1.4.0

### ISSUE THREAD

**Issue by QuanluZhang** [pytorch/pytorch#32641]
## ❓ Questions

For the following example code, the printed graph in torch 1.4.0 is very different from that in torch 1.3.1. I am wondering is this expected? because many useful fields are removed in torch 1.4.0.

```
import torch
import torch.nn as nn

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv = nn.Conv2d(1, 1, 3)

    def forward(self, x):
        return self.conv(x)

if __name__ == '__main__':
    n = Net()
    example_weight = torch.rand(1, 1, 3, 3)
    example_forward_input = torch.rand(1, 1, 3, 3)

    # Trace a specific method and construct `ScriptModule` with
    # a single `forward` method
    module = torch.jit.trace(n.forward, example_forward_input)

    # Trace a module (implicitly traces `forward`) and construct a
    # `ScriptModule` with a single `forward` method
    module = torch.jit.trace(n, example_forward_input)
    print(module.graph)
```
output in torch 1.3.1:
```
graph(%self : ClassType<Net>,
      %input : Float(1, 1, 3, 3)):
  %1 : ClassType<Conv2d> = prim::GetAttr[name="conv"](%self)
  %weight : Tensor = prim::GetAttr[name="weight"](%1)
  %3 : Tensor = prim::GetAttr[name="bias"](%1)
  %5 : int = prim::Constant[value=1](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %6 : int = prim::Constant[value=1](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %7 : int[] = prim::ListConstruct(%5, %6), scope: Net/Conv2d[conv]
  %8 : int = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %9 : int = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %10 : int[] = prim::ListConstruct(%8, %9), scope: Net/Conv2d[conv]
  %11 : int = prim::Constant[value=1](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %12 : int = prim::Constant[value=1](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %13 : int[] = prim::ListConstruct(%11, %12), scope: Net/Conv2d[conv]
  %14 : bool = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %15 : int = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %16 : int = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %17 : int[] = prim::ListConstruct(%15, %16), scope: Net/Conv2d[conv]
  %18 : int = prim::Constant[value=1](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %19 : bool = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %20 : bool = prim::Constant[value=0](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %21 : bool = prim::Constant[value=1](), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  %22 : Float(1, 1, 1, 1) = aten::_convolution(%input, %weight, %3, %7, %10, %13, %14, %17, %18, %19, %20, %21), scope: Net/Conv2d[conv] # /home/xxx/.local/lib/python3.5/site-packages/torch/nn/modules/conv.py:342:0
  return (%22)
```

output in torch 1.4.0:
```
graph(%self.1 : __torch__.torch.nn.modules.module.___torch_mangle_4.Module,
      %input : Float(1, 1, 3, 3)):
  %28 : __torch__.torch.nn.modules.module.___torch_mangle_3.Module = prim::GetAttr[name="conv"](%self.1)
  %30 : Tensor = prim::CallMethod[name="forward"](%28, %input)
  return (%30)
```

**Comment by suo** [comment]
Yes, this is expected. We changed the internal representation of the graph to preserve call structure. This makes some graph transformations easier, because they may rely on the call graph. We consider the graph largely an internal detail so we didn't mention it in the release notes.

If you want a fully inlined graph, you can invoke our inlining pass:
```
torch._C._jit_pass_inline(graph)
```

We may add a quick `inlined_graph` binding to `Function` to make this a little easier.

**Comment by QuanluZhang** [comment]
@suo thanks for your quick reply. `torch._C._jit_pass_inline(graph)` works, it can output the complete graph.

But the scope field is different from that in torch 1.3.1. In torch 1.3.1 every graph node has the scope field, but in the inlined graph the graph nodes in `forward` (such as `+` `avgpool`) do not have the scope field.
Moreover, the format of the scope field is also different, I can get module type (e.g., `Conv2d`) from the scope field in torch 1.3.1, but in 1.4.0 the type information is removed. Just wondering is it possible to get the graph format the same as in torch 1.3.1?

**Comment by suo** [comment]
Sadly, we cannot output exactly the same graph format as torch 1.3.1. The graph is considered an internal detail and thus is subject to change; the JIT only guarantees backward compatibility for saved ScriptModules.

May I ask what you're using the scope for? Scopes have always been a little undermaintained, and we are thinking about exposing alternatives that we expect to be more stable.

**Comment by QuanluZhang** [comment]
@suo , i see, thanks.

Here is our scenario. we are working on model pruning, usually model pruning is done by inserting masks on weights. This approach can check the accuracy of a model when applying specific values of masks, but it does not show really speedup on model inference, it is just a simulation. We are trying to automatically convert the original model to a small model based on the masks. To do this we need a graph representation of the model, but `named_modules` only has graph nodes, there is no connection between them. Thus we tried `jit.trace`, it has rich information, e.g., connections, operation type, module name, tensor shape.

It would be great if you can tell me other alternatives for getting those information of a model.

**Comment by jamesr66a** [comment]
cc @raghuramank100 model pruning

**Comment by chenbohua3** [comment]
I think the graph representation in Pytorch 1.3.1 is better for secondary development. For example, we are going to directly convert Pytorch module to TensorRT without ONNX. In Pytorch 1.3.1 we can use the inter-connections of the nodes in the graph and Depth First Search to construct maps from Pytorch graph to TensorRT graph. But in Pytorch 1.4.0, the relationship between nodes is hidden.

For the module described below:
```
class Test(nn.Module):
    def __init__(self):
        super(Test, self).__init__()
        self.fc1 = nn.Linear(1, 2)

    def subCall(self, x):
        return self.fc1(x)

    def forward(self, x):
        return self.subCall(x)
```
How can we parse self.fc1 from the graph generated by `jit.script`:
```
graph(%self : __torch__.Test,
      %x.1 : Tensor):
      %3 : Tensor = prim::CallMethod[name="subCall"](%self, %x.1) 
      return (%3)
```

**Comment by suo** [comment]
@chenbohua3 if you want the full inlined graph, you can call our inlining pass on it: `torch._C._jit_pass_inline`.

**Comment by suo** [comment]
> It would be great if you can tell me other alternatives for getting those information of a model.

@QuanluZhang: so beyond inlining, if you want to recover the call stack (basically get the same information that scopes were giving you), try using `Node::callstack()`. This should give you a data structure with the info you're looking at

**Comment by chenbohua3** [comment]
@suo It works for me, thank you:) 
By the way, do you have roadmaps for JIT? We are evaluating various deployment approaches of Pytorch model,  roadmaps can help us make better decisions.

**Comment by suo** [comment]
@chenbohua3 we don't publish overall roadmaps, but generally when we work on large features or improvements we'll post an RFC issue.

**Comment by xvdp** [comment]
This is a bummer, a setback.
I just put 1.4 on one of my machines and found this out.  Fully exposing the graph can be useful for research into automl. This will force me to have to rewrite my graphs with a generator to have a trace of them them on creation. I was hoping that 1.4 would expose the onnx instead of hiding it.

**Comment by xvdp** [comment]
I see, Ill build an onnx importer and work from that instead of going thru the graph as i was before.

**Comment by xvdp** [comment]
Ok,parsing onnx graph is pretty clean. It would be nice to have the protobuf data before the export, ie. the step just before torch.onnx.export to return the object without writing to file.

**Comment by suo** [comment]
@xvdp, does `torch._C._jit_pass_inline` work for you to retrieve the fully inlined graph?

**Comment by xvdp** [comment]
just a bit of a pain, but looks like it should, you  mean, used like this...?
```python
trace = torch.jit.trace(net, image)
torch._C._jit_pass_inline(trace.graph)
it = trace.graph.nodes().__iter__()
this = next(it)
print(this.scopeName(), this.kind(), list(this.inputs()), list(this.outputs()))
```
thanks.

**Comment by suo** [comment]
yeah. https://github.com/pytorch/pytorch/pull/33508 adds an `inlined_graph` property to make this more convenient (so instead of calling `trace.graph` you could call `trace.inlined_graph`.

**Comment by xvdp** [comment]
I may have found a cleaner way of doing it.
I noticed that you guys ship with code that parses this so I can just leverge that pytorch/pytorch/torch/utils/tensorboard/_pytorch_graph.py

```python
import torch.utils.tensorboard._pytorch_graph as pg
g = pg.graph(net, image)
then iterate thru g[0].node
```

**Comment by ZizhouJia** [comment]
torch._C._jit_pass_inline(trace.graph) is not work for ONNX jit graph to expose the scopeName. 
Could you tell me How to expose the scopeName in ONNX jit graph.
The scopeName is very usefully. I also was hoping that it would better expose the onnx instead of hiding it.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
